### PR TITLE
Ensure points are summed with full precision

### PR DESCRIPTION
**Pull request by nwalters512**
Closes https://github.com/PrairieLearn/PrairieLearn/issues/10928.

I left comments everywhere since this is not very obvious and since we definitely want to remember to handle this if/when we convert these things to TypeScript.

I didn't originally identify the need to change `instance_question_points_homework` in #10928. However, it came up in my grep, and I was able to reproduce a "failure" that involved it:

- Update the `points` of the first question in `testCourse/courseInstances/Sp15/assessments/hw3-partialCredit/infoAssessment.json` to `0.1`
- Start the assessment as a student.
- Make 3 100% submissions to the first question.
- Open the assessment log page and click the pencil to edit the points on the first question.
- Observe that the value is `0.3000000000...`.

After this change, that's correctly computed as `0.3` after following the same steps.

I also tested the original issue. I updated the same `testCourse/courseInstances/Sp15/assessments/hw3-partialCredit/infoAssessment.json` to have `"
type": "Exam"` and updated the questions to the following:

```json
{
  "zones": [
    {
      "questions": [
        { "id": "partialCredit1", "points": [2] },
        { "id": "partialCredit2", "points": [2] },
        { "id": "partialCredit3", "points": [1.6] },
        { "id": "partialCredit4_v2", "points": [1.6] }
      ]
    }
  ]
}
```

I then started the assessment as a student, submitted 100% to everything, and then checked the score on the assessment instance log page:

Screenshot 2024-11-15 at 16 32 24

After making these changes, both `points` and `max_points` are correctly computed as 7.2.

I've focused exclusively on sums in this PR. It's possible that there's also issues with how we use multiplication, e.g. to compute percentages? I didn't observe any issues in my testing that weren't fixed by ensuring that points are always computed exactly.

### CODE REVIEW

File: apps/prairielearn/src/sprocs/instance_questions_points_homework.sql [hunk 1]
```diff
@@ -69,7 +69,9 @@ BEGIN
     length := cardinality(variants_points_list);
     auto_points := 0;
     FOR i in 1..length LOOP
-        auto_points := auto_points + variants_points_list[i];
+        -- We convert values to `numeric` to avoid accumulating floating-point
+        -- errors when summing non-integer points.
+        auto_points := auto_points::numeric + variants_points_list[i]::numeric;
```
Comment (mwest1066) [inline apps/prairielearn/src/sprocs/instance_questions_points_homework.sql:74]: This is fine but I think it would be even safer to declare a new `auto_points_accumulator` variable that has type `numeric` for use in the loop, then convert back to the double-precision `auto_points` only once at the end. At the moment we are forcing the accumulator back to double-precision at each loop iteration.

**Review by mwest1066 — APPROVED** [review]

**Comment by nwalters512** [comment]
I worry that this problem goes deeper than this PR. Here's another case that fails (this is a Homework):

```json
{
  "zones": [
    {
      "questions": [
        {
          "id": "partialCredit1",
          "autoPoints": 0.3,
          "constantQuestionValue": true,
          "maxAutoPoints": 1,
          "manualPoints": 0.1
        }
      ]
    }
  ]
}
```

We end up computing the question value as `0.30000000000000004`, which is displayed to students:

Screenshot 2024-11-18 at 09 34 44

This value is computed in JavaScript, so we can't change this with Postgres type conversions, we'd have to do something like what @mwest1066 described in https://github.com/PrairieLearn/PrairieLearn/issues/10928#issuecomment-2480033630:

https://github.com/PrairieLearn/PrairieLearn/blob/425082392f0a774d119d6ab356bded6d0c6b2351/apps/prairielearn/src/components/QuestionScore.html.ts#L479-L480

We actually end up with `{0.30000000000000004}` in `instance_questions.variants_points_list` because of this calculation (which in this example does `0.4 - 0.1`, same as above:

https://github.com/PrairieLearn/PrairieLearn/blob/425082392f0a774d119d6ab356bded6d0c6b2351/apps/prairielearn/src/sprocs/instance_questions_points_homework.sql#L44

Even given that, the changes to `instance_questions_points_homework` mean the _final_ score is still computed correctly, because of this fact:

```sql
SELECT (0.30000000000000004::double precision)::numeric;
> 0.3
```

I tend to think that we should always use exact math for points values, though I also think that it's probably not worth it to try to make things perfect in this PR. Thoughts?

File: apps/prairielearn/src/sprocs/instance_questions_points_homework.sql (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/sprocs/instance_questions_points_homework.sql]: I made this change in 9f3a619, how's that look?

**Comment by mwest1066** [comment]
> I've focused exclusively on sums in this PR. It's possible that there's also issues with how we use multiplication, e.g. to compute percentages? I didn't observe any issues in my testing that weren't fixed by ensuring that points are always computed exactly.

Here's 30% of 8.2 (should be 2.56):
```
node
> 8.2 * 30 / 100
2.4599999999999995
```

Two alternative approaches for this PR are listed below. We could certain still merge this PR even if we want to proceed with one of the alternatives below, or we could solely use of them instead.

## Alternative #1: DP calculations and storage, round on display

We could do all computations in double precision, but always round values to 6 decimal places before displaying or exporting them.

For example, when displaying a `score_perc` we would do something like:
```
score_perc = 3.293591391942314;
console.log(`exact value stored in DB: ${score_perc}`);
console.log(`value shown to students: ${score_perc.toFixed(6).slice(null, -5)}`);
console.log(`value exported in CSV or LTI: ${score_perc.toFixed(6)}`);
```

This renders 79.9999999 as 80, and 30.98 as 30.9, which is what we want in both cases.

This alternative is a relatively easy change to make because we only need to change our display routines. We'd secretly have more precision, but it's never going to practically matter to anyone. The only real downside I can think of is that we would in fact be hiding data that we actually have, but I don't think that anyone is ever going to care. We could be clear about this in the docs ("all values are stored and computed in double precision, but are rounded to 6 decimal places before viewing or exporting").

## Alternative #2: store in fixed point, round on store

We could store everything as integer micro-points (one millionth of a point). That is, rather than storing `points: double precision` we'd store `micro_points: integer`. When we do calculations, we perform operations in double precision but then round (not floor) when computing micro-points:
```
iq.micro_points = round(gj.score * aq.points * 1e6)

ai.micro_points = sum(iq.micro_points)
```
In the above calculation everything is integers. In postgres these would be actual integers, in TS they'd be integers stored as Numbers. Because integers are exactly represented in Number (double precision) all sums would be exact.

When we want to display a value we'd still floor the displayed value, like:
```
micro_score_perc = 3293591.391942314;
console.log(`value shown to students: ${(micro_score_perc / 1e6).toFixed(8).slice(null, -7)}`);
// prints 3.2
```

To migrate to the new columns, probably the best strategy would be to add the new columns, start double-writing, backfill the new columns, and then switch the the new ones.

This alternative has the advantage that we could precisely show exactly what our intermediate storage values are, and all of our summing operations would be exact. The downside is that the change would be quite invasive.

**Comment by mwest1066** [comment]
> I worry that this problem goes deeper than this PR.

Oops, sorry, I wrote the [comment above](https://github.com/PrairieLearn/PrairieLearn/pull/10930#issuecomment-2483986355) while offline yesterday and I hadn't seen [your comment](https://github.com/PrairieLearn/PrairieLearn/pull/10930#issuecomment-2483734504) yet. I agree that this is all tricker than just summing up values.

**Comment by nwalters512** [comment]
I think alternative #2 is invasive enough that I want to completely discard it. That makes every bit of math/display/etc. of points incredibly annoying and it'd almost certainly be really hard to do correctly.

Alternative #1 is reasonable. It has the same downside of requiring careful work every time we display/output things, though it's better in that we don't have to dramatically change how we do calculations themselves.

A third option could be to take this PR to its logical conclusion: we can always convert points to `numeric` before doing calculations, then convert back to floating point numbers in the end (whatever that might be: either before returning into JS, or before storing back into the database. As I pointed out above, there are cases where we do point math in JavaScript, so we'd probably have to reach for something like https://www.npmjs.com/package/decimal.js or implement some helpers ourselves.

I don't have a great sense of how difficult these two options would be relative to each other. I think I lean towards "just do the math right" as the technically correct thing to do here and the one that'll probably be the easiest to maintain in the long run?

**Comment by jonatanschroeder** [comment]
In my view there are basically three kinds of cases:
* Cases where we round to a set number of digits. This should not be a problem in general, provided we use a proper method of rounding (something more sophisticated than `Math.round(val * 100) / 100`).
* Cases where we floor to a set number of digits. The issue in this case is that values that are very close to a round number floor down too much. One way to handle this is to consider a particular grade that we assume will never be significant (e.g., 1e-6), and add it to the value before flooring (e.g., always use `score+1e-6` as the argument to floor).
* Cases where we don't round or floor, and we show the full number. We probably want to revisit these cases and always format them with a maximum set of digits.

**Comment by mwest1066** [comment]
I don't necessarily object to "do the math right" with an arbitrary-precision library, although I don't love the idea of having code on the hot path that can do all kinds of crazy allocations to maintain exact precision. And I hope we don't have any divisions, because they can really mess things up.

For consideration, yet another alternative:

## Alternative #3: Store and calculate in DP, round before store

This is kind of a hybrid of alternatives 1 and 2 above, but without the invasive nature of 2.

Basically, we'd continue to store and calculate exactly like we do right now, using double precision everywhere. But whenever we store a value back to the DB we'd round to 6 decimal places.

Unfortunately, the only easy way I know of to do this is:
```
stored_value = Number(value.toFixed(6));
```
This is a bit janky, but it has the behavior that we want.

We can then continue to display as we do now.

**Comment by mwest1066** [comment]
And just to help when discussing:

## Alternative #4: just do the math right

Use `numeric` type in postgres and something like [decimal.js](https://www.npmjs.com/package/decimal.js) in TS to perform all calculations to arbitrary precision, before rounding to store as DP.

This should generally work fine in terms of giving the correct answer, because these methods round on conversion _from_ DP as well as _to_ DP. I'm not sure how well they handle the case of division, but we might not do any divisions so this could be irrelevant.

**Comment by reteps** [comment]
I think Alternative #1 is the most reasonable. Alternative #2 I am not a fan of, alternative #3 feels like it could lead to strange situations if we are reading values from the DB, calculating, and writing them back. Alternative #4, feels the best, but would require us to move all score logic to TS? I would want to do that eventually for ease of maintenance, but it doesn't feel high priority.

**Comment by mwest1066** [comment]
I agree that Alternative #1 continues to be my preferred solution for now.

We definitely want to move score logic to TS because we'll need this before rewriting bits of it (e.g., to support #3347). But even if we implement Alternative #4 in the future, it would be good to do Alternative #1 right away. These are complementary because even with exact score calculations (Alternative #4) we still don't want to display more than 6 decimal places of a score (Alternative #1).

**Comment by nwalters512** [comment]
I agree that option 1 will be our best bet. I wrote down that conclusion in https://github.com/PrairieLearn/PrairieLearn/issues/10928. I'll close out this PR since I don't think there's anything we want to save beyond the discussion.
